next.middleware.ts

```diff
@@ -0,0 +1,39 @@
+import { availableLocales, getCurrentLocale } from './next.locales.mjs';
+import type { NextRequest, NextResponse } from 'next/server';
+import type { CustomMiddleware, NextMiddlewareLocale } from './types';
+
+const availableLocaleCodes = availableLocales.map(locale => locale.code);
+
+const createMiddleware = (response: typeof NextResponse) => {
+  const localeMiddleware = (request: NextRequest): NextMiddlewareLocale => {
+    const currentLocale = getCurrentLocale(request.nextUrl.pathname);
+
+    return {
+      currentLocale: currentLocale.code,
+      availableLocales: availableLocaleCodes,
+      isLocalisedRoute: currentLocale.code !== 'en',
+    };
+  };
+
+  return (middlewares: CustomMiddleware[]) => {
+    const middleware = async (request: NextRequest) => {
+      const locale = localeMiddleware(request);
+
+      const matchedMiddleware = middlewares.find(middleware =>
+        middleware.matcher(request, locale)
```

Comment:
I think that we don't need to pass `localeMiddleware` here. The request can be parsed inside the `detectLanguage` middleware itself.

IMO middleware function needs to be generic (only Request & Response) otherwise it feels too specific.

Reply:
The middleware matcher is a function that requires current context. i18n is a default Next.js context. When using built-in i18n functionality, you can see that the whole locale data is available there, which makes sense. Since i18n is a fundamental part of routing when enabled.

Hence, I don't see any issues having locale data here. At best, it can be undefined, which is fine.

Reply:
@HinataKah0 I see that your review is a review comment and not a requested changes. I'm going then proceed with merging this PR. If you still feel further changes are needed feel free to reply here or Slack and we can make a follow-up conversation 🙇

Reply:
@HinataKah0 I see that your review is a review comment and not a requested changes. I'm going then proceed with merging this PR. If you still feel further changes are needed feel free to reply here or Slack and we can make a follow-up conversation 🙇

Reply:
Got it!
Thanks for explaining!